internal/status/ing_status.go

```
@@ -105,28 +115,155 @@ func updateObject(mIngress *networkingv1beta1.Ingress, updateOption UpdateStatus
+			utils.AviLog.Errorf("key: %s, error in unmarshalling Ingress %s/%s annotations for VS: %v",
+				key, ingObj.Namespace, ingObj.Name, err)
+		}
+	}
+
+	// update the existing hostname vs uuid for the current update
+	for i := 0; i < len(svcMetaHostnames); i++ {
+		vsAnnotations[svcMetaHostnames[i]] = vsUUID
+	}
+
+	// remove the hostname from annotations which is not part of the spec
+	for k := range vsAnnotations {
+		if !utils.HasElem(ingHostList, k) {
+			delete(vsAnnotations, k)
+		}
+	}
+
+	// compare the vs annotations for this ingress object
+	req := isAnnotationsUpdateRequired(ingObj.Annotations, vsAnnotations)
+	if !req {
+		utils.AviLog.Debugf("annotations update not required for this ingress: %s/%s", ingObj.Namespace, ingObj.Name)
+		return nil
+	}
+	if err = patchIngressAnnotations(ingObj, vsAnnotations, mClient); err != nil {
+		utils.AviLog.Errorf("key: %s, msg: there was an error in updating the ingress annotations: %v", key, err)
 		// fetch updated ingress and feed for update status
-		mIngresses := getIngresses([]string{mIngress.Namespace + "/" + mIngress.Name}, false)
+		mIngresses := getIngresses([]string{ingObj.Namespace + "/" + ingObj.Name}, false)
 		if len(mIngresses) > 0 {
-			return updateObject(mIngresses[mIngress.Namespace+"/"+mIngress.Name], updateOption, retry+1)
+			return updateIngAnnotations(mClient, mIngresses[ingObj.Namespace+"/"+ingObj.Name], svcMetaHostnames, vsUUID, key, ingHostList, oldIng, retry+1)
+		}
+	}
+
+	return nil
+}
+
+func isAnnotationsUpdateRequired(ingAnnotations map[string]string, newVSAnnotations map[string]string) bool {
+	oldVSAnnotationsStr, ok := ingAnnotations[VSAnnotation]
+	if !ok {
+		if len(newVSAnnotations) > 0 {
+			return true
+		}
+		return false
+	}
+
+	var oldVSAnnotations map[string]string
+	if err := json.Unmarshal([]byte(oldVSAnnotationsStr), &oldVSAnnotations); err != nil {
+		utils.AviLog.Errorf("error in unmarshalling old vs annotations %s: %v", oldVSAnnotationsStr, err)
```

If you weren't able to unmarshall the annotations, could there be a bigger problem here and you fail the update instead of returning true?

---

The annotations wouldn't be in proper format in this case, and hence, it would be better to return true and update the annotations I think.